# Fake News Classifier

###### Importing required libraries

In [40]:
import numpy as np
import pandas as pd

###### Reading data from csv files 

In [14]:
df1 = pd.read_csv('true.csv')
df2 = pd.read_csv('fake.csv')
df1['label']='TRUE'
df2['label']='FAKE'

In [12]:
print(df1.shape)
df1.head()

(21417, 5)


title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date label  
0  December 31, 2017   TRUE  
1  December 29, 2017   TRUE  
2  December 31, 2017   TRUE  
3  December 30, 2017   TRUE  
4  December 29, 2017   TRUE

In [15]:
print(df2.shape)
df2.head()

(23481, 5)


title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date label  
0  December 31, 2017  FAKE  
1  December 31, 2017  FAKE  
2  December 30, 2017  FAKE  
3  December 29, 2017  FAKE  
4  December 25, 2017  FAKE

###### Combining the two datasets containing Real and Fake News into one dataset

In [41]:
df = pd.concat([df1, df2])

In [17]:
print(df.shape)
df.describe()

(44898, 5)


title   text  \
count                                               44898  44898   
unique                                              38729  38646   
top     Factbox: Trump fills top jobs for his administ...          
freq                                                   14    627   

             subject                date  label  
count          44898               44898  44898  
unique             8                2397      2  
top     politicsNews  December 20, 2017    FAKE  
freq           11272                 182  23481

In [18]:
df.info

<bound method DataFrame.info of                                                    title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuter

###### Cleaning the data by dropping columns that aren't required

In [21]:
df.drop(['date','subject'],inplace=True,axis =1)

In [22]:
df.describe

<bound method NDFrame.describe of                                                    title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text label  
0      WASHINGTON (Reuters) - The head of a conservat...  TRUE  
1      WASHINGTON (Reuters) - Transgender people will...  TRUE  
2      WASHINGTON (Reuters) - The special counsel 

###### As we concatenated two datasets, we need to shuffle Real and Fake data

In [42]:
df= df.sample(frac=1)

In [24]:
df.describe

<bound method NDFrame.describe of                                                    title  \
13560  LIST OF 15 CORPORATIONS Working With Obama To ...   
6793   CIA says Russia intervened to help Trump win W...   
8327   Mexico says president Pena Nieto to meet Trump...   
15689  RECKLESS DEM MAYOR BLAMES AMTRAK ENGINEER FOR ...   
5304   California lawmakers release environmental bil...   
...                                                  ...   
1707    WATCH: Republican Heartlessly Tells A Mother ...   
13150  MIKE ROWE SENDS A BRUTAL MESSAGE To The Media ...   
715    U.S. Senate confirms top EPA air quality offic...   
9395   Obama administration officials warn lawmakers ...   
7522   Monmouth poll: 6-point Clinton lead over Trump...   

                                                    text label  
13560  Buckle up America Obama still has 6 months to ...  FAKE  
6793   WASHINGTON (Reuters) - The CIA has concluded t...  TRUE  
8327   MEXICO CITY (Reuters) - Mexican President E

In [25]:
labels = df.label
labels.head()

13560    FAKE
6793     TRUE
8327     TRUE
15689    FAKE
5304     TRUE
Name: label, dtype: object

In [26]:
df.head()

title  \
13560  LIST OF 15 CORPORATIONS Working With Obama To ...   
6793   CIA says Russia intervened to help Trump win W...   
8327   Mexico says president Pena Nieto to meet Trump...   
15689  RECKLESS DEM MAYOR BLAMES AMTRAK ENGINEER FOR ...   
5304   California lawmakers release environmental bil...   

                                                    text label  
13560  Buckle up America Obama still has 6 months to ...  FAKE  
6793   WASHINGTON (Reuters) - The CIA has concluded t...  TRUE  
8327   MEXICO CITY (Reuters) - Mexican President Enri...  TRUE  
15689  Perhaps Democrat Mayor Michael Nutter is takin...  FAKE  
5304   SAN FRANCISCO (Reuters) - Democratic state sen...  TRUE

###### Importing required libraries

In [43]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier

###### Splitting the data into training and test set

In [29]:
X_train, X_test, y_train, y_test = train_test_split(df.text,labels, test_size=0.2, random_state =5)

###### Initialise Tdidf vectorizer with stopwords and fit the vectorizer with training data and transform it using test data

In [31]:
tfidfv = TfidfVectorizer(stop_words = 'english', max_df =0.5)
tfidf_train = tfidfv.fit_transform(X_train)
tfidf_test = tfidfv.transform(X_test)

In [34]:
classifier = PassiveAggressiveClassifier(max_iter = 50)

In [35]:
classifier.fit(tfidf_train, y_train)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

In [36]:
y_pred = classifier.predict(tfidf_test)

In [37]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

###### Measuring Accuracy score of the model

In [38]:
accuracy = accuracy_score(y_test,y_pred)
percentAccuracy= round(accuracy*100,2)
print(f'{percentAccuracy}%')

99.33%


> The percentage accuracy of this model is 99.33%

###### Making a confusion matrix to analyse and compare teh given and predicted values

In [39]:
cm = confusion_matrix(y_test,y_pred)
print(cm)

[[4726   33]
 [  27 4194]]


> This model gives 4726 true positives, 4194 true negatives, 27 false positives and 33 false negatives. 